File: lib/models/pelanggan.dart
import '../utils/numeric_parser.dart';

class Pelanggan {
  final String id;
  final String nama;
  final String noHp;
  final String? alamat;
  final int totalGalonPinjam;
  final double totalTransaksi;
  final String? catatan;
  final bool isAktif;
  final DateTime createdAt;
  final DateTime? updatedAt;

  const Pelanggan({
    required this.id,
    required this.nama,
    required this.noHp,
    this.alamat,
    this.totalGalonPinjam = 0,
    this.totalTransaksi = 0,
    this.catatan,
    this.isAktif = true,
    required this.createdAt,
    this.updatedAt,
  });

  factory Pelanggan.fromJson(Map<String, dynamic> json) => Pelanggan(
        id: json['id'] as String,
        nama: json['nama'] as String,
        noHp: json['noHp'] as String? ?? '',
        alamat: json['alamat'] as String?,
        totalGalonPinjam: parseInt(json['totalGalonPinjam']),
        totalTransaksi: parseDouble(json['totalTransaksi']),
        catatan: json['catatan'] as String?,
        isAktif: json['isAktif'] as bool? ?? true,
        createdAt: DateTime.parse(json['createdAt'] as String),
        updatedAt: json['updatedAt'] != null
            ? DateTime.parse(json['updatedAt'] as String)
            : null,
      );

  Map<String, dynamic> toJson() => {
        'id': id,
        'nama': nama,
        'noHp': noHp,
        'alamat': alamat,
        'totalGalonPinjam': totalGalonPinjam,
        'totalTransaksi': totalTransaksi,
        'catatan': catatan,
        'isAktif': isAktif,
        'createdAt': createdAt.toIso8601String(),
        'updatedAt': updatedAt?.toIso8601String(),
      };
}

File: backend_laravel/app/Models/Pelanggan.php
<?php

namespace App\Models;

use Illuminate\Database\Eloquent\Model;
use Illuminate\Database\Eloquent\Relations\HasMany;

class Pelanggan extends Model
{
    protected $table = 'pelanggan';

    public $incrementing = false;

    protected $keyType = 'string';

    const UPDATED_AT = null;

    protected $fillable = [
        'id',
        'nama',
        'no_hp',
        'alamat',
        'total_galon_pinjam',
        'total_transaksi',
        'catatan',
        'is_aktif',
    ];

    protected $casts = [
        'total_galon_pinjam' => 'integer',
        'total_transaksi' => 'decimal:2',
        'is_aktif' => 'boolean',
        'created_at' => 'datetime',
    ];

    public function galon(): HasMany
    {
        return $this->hasMany(Galon::class, 'pelanggan_id');
    }

    public function transaksi(): HasMany
    {
        return $this->hasMany(Transaksi::class, 'pelanggan_id');
    }

    public function mutasiGalon(): HasMany
    {
        return $this->hasMany(GalonMutasi::class, 'pelanggan_id');
    }
}
